File: ESimManager/Models/ESimProfile.cs
namespace ESimManager.Models;

/// <summary>
/// Represents an eSIM profile with GSMA SGP.22 compliant metadata
/// </summary>
public class ESimProfile
{
    public int Id { get; set; }
    
    /// <summary>
    /// Integrated Circuit Card Identifier - unique 19-20 digit identifier
    /// </summary>
    public string Iccid { get; set; } = string.Empty;
    
    public string Name { get; set; } = string.Empty;
    
    public string ProviderName { get; set; } = string.Empty;
    
    /// <summary>
    /// Current state of the profile (Disabled, Enabled, Deleted)
    /// </summary>
    public ProfileState State { get; set; }
    
    /// <summary>
    /// Profile classification (Operational, Test, Provisioning)
    /// </summary>
    public ProfileClass ProfileClass { get; set; }
    
    public DateTime? ActivatedDate { get; set; }
    
    public DateTime? DeactivatedDate { get; set; }
    
    public DateTime CreatedDate { get; set; }
    
    public DateTime ModifiedDate { get; set; }
    
    public string CustomNotes { get; set; } = string.Empty;
    
    /// <summary>
    /// Encrypted activation code stored at rest using DPAPI
    /// </summary>
    public byte[]? EncryptedActivationCode { get; set; }
    
    /// <summary>
    /// Foreign key to associated device
    /// </summary>
    public int? DeviceId { get; set; }
}

/// <summary>
/// GSMA SGP.22 compliant profile states
/// </summary>
public enum ProfileState
{
    Disabled,
    Enabled,
    Deleted
}

/// <summary>
/// Profile classification per GSMA standards
/// </summary>
public enum ProfileClass
{
    Operational,
    Test,
    Provisioning
}

/// <summary>
/// Legacy status enum for backward compatibility
/// </summary>
public enum ESimStatus
{
    Inactive,
    Active,
    Disabled,
    Error
}

File: api/src/models/ESimProfile.ts
import mongoose, { Document, Schema } from 'mongoose'

export type ProfileStatus = 'active' | 'inactive' | 'pending' | 'deleted'
export type ProfileClass = 'operational' | 'test' | 'provisioning'

export interface IESimProfile extends Document {
  iccid: string
  name: string
  provider: string
  status: ProfileStatus
  profileClass: ProfileClass
  userId: mongoose.Types.ObjectId
  deviceId?: mongoose.Types.ObjectId
  activatedDate?: Date
  deactivatedDate?: Date
  customNotes?: string
  metadata?: Record<string, any>
  createdAt: Date
  updatedAt: Date
}

const esimProfileSchema = new Schema<IESimProfile>({
  iccid: { type: String, required: true, unique: true },
  name: { type: String, required: true },
  provider: { type: String, default: 'NexoraSIM' },
  status: { type: String, enum: ['active', 'inactive', 'pending', 'deleted'], default: 'pending' },
  profileClass: { type: String, enum: ['operational', 'test', 'provisioning'], default: 'operational' },
  userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  deviceId: { type: Schema.Types.ObjectId, ref: 'Device' },
  activatedDate: { type: Date },
  deactivatedDate: { type: Date },
  customNotes: { type: String },
  metadata: { type: Schema.Types.Mixed }
}, { timestamps: true })

esimProfileSchema.index({ userId: 1 })
esimProfileSchema.index({ status: 1 })
esimProfileSchema.index({ iccid: 1 })

export const ESimProfile = mongoose.model<IESimProfile>('ESimProfile', esimProfileSchema)
